File: ios/Sources/LivelineKit/Math/MomentumDetect.swift
import Foundation

/// Auto-detects the line's ``Trend`` from recent samples, matching web
/// liveline's `detectMomentum`.
///
/// Only triggers during active movement: the threshold is a fraction of the
/// full-lookback range, but the velocity is measured over just the last few
/// points, so a line that drifted long ago but is now steady reads as `flat`.
public enum MomentumDetect {
    /// Detects the trend.
    /// - Parameters:
    ///   - points: The series (ascending in time).
    ///   - lookback: How many trailing points define the range (default 20).
    public static func detect(_ points: [LivelinePoint], lookback: Int = 20) -> Trend {
        guard points.count >= 5 else { return .flat }
        let start = Swift.max(0, points.count - lookback)

        var minV = Double.infinity
        var maxV = -Double.infinity
        for i in start..<points.count {
            let v = points[i].value
            if v < minV { minV = v }
            if v > maxV { maxV = v }
        }
        let range = maxV - minV
        if range == 0 { return .flat }

        let tailStart = Swift.max(start, points.count - 5)
        let first = points[tailStart].value
        let last = points[points.count - 1].value
        let delta = last - first
        let threshold = range * 0.12

        if delta > threshold { return .up }
        if delta < -threshold { return .down }
        return .flat
    }
}

File: android/liveline-core/src/main/kotlin/com/liveline/core/MomentumDetect.kt
package com.liveline.core

import kotlin.math.max

/**
 * Auto-detects the line's [Trend] from recent samples, matching web liveline's
 * `detectMomentum`.
 *
 * Only triggers during active movement: the threshold is a fraction of the
 * full-lookback range, but the velocity is measured over just the last few
 * points, so a line that drifted long ago but is now steady reads as `flat`.
 */
object MomentumDetect {
    /**
     * Detects the trend.
     * @param points The series (ascending in time).
     * @param lookback How many trailing points define the range (default 20).
     */
    fun detect(points: List<LivelinePoint>, lookback: Int = 20): Trend {
        if (points.size < 5) return Trend.FLAT
        val start = max(0, points.size - lookback)

        var minV = Double.POSITIVE_INFINITY
        var maxV = Double.NEGATIVE_INFINITY
        for (i in start until points.size) {
            val v = points[i].value
            if (v < minV) minV = v
            if (v > maxV) maxV = v
        }
        val range = maxV - minV
        if (range == 0.0) return Trend.FLAT

        val tailStart = max(start, points.size - 5)
        val first = points[tailStart].value
        val last = points[points.size - 1].value
        val delta = last - first
        val threshold = range * 0.12

        return when {
            delta > threshold -> Trend.UP
            delta < -threshold -> Trend.DOWN
            else -> Trend.FLAT
        }
    }
}
